### PR TITLE
[CORE + NOT TO MERGE] Changing the Vector to BoundedVector (Idem with matrices) in ConstitutiveLaws

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR we will be adding and discussing the substitution of the Vector/Matrix types with BoundedVector/BoundedMAtrix with the resize option. 
I've only added a few to start the discussion, 

I have doubts regarding the      mpShapeFunctionsValues and the  mpShapeFunctionsDerivatives for example...

### CODE REVIEW

File: kratos/includes/constitutive_law.h
Comment (loumalouomega) [inline kratos/includes/constitutive_law.h]: Fo the moment this is OK, the pain will be actually replace the laws

Comment (loumalouomega) [reply kratos/includes/constitutive_law.h]: Are we going to consider always 3D by default?

Comment (loumalouomega) [inline kratos/includes/constitutive_law.h]:
```suggestion
    typedef array_1d<double, 6> StrainVector3DType;
```

File: kratos/includes/constitutive_law.h
Comment (loumalouomega) [inline kratos/includes/constitutive_law.h]:
```suggestion
    typedef array_1d<double, 6> StressVector3DType;
```

File: kratos/includes/constitutive_law.h
Comment (loumalouomega) [inline kratos/includes/constitutive_law.h]:
```suggestion
    typedef BoundedMatrix<double, 6, 6> VoigtSizeMatrix3DType;     // Constitutive Matrix
```

File: kratos/includes/constitutive_law.h
Comment (loumalouomega) [inline kratos/includes/constitutive_law.h]:
```suggestion
    typedef BoundedMatrix<double, 3, 3> DimensionSizeMatrix3DType; // Def. gradient tensor
```

File: kratos/includes/constitutive_law.h
Comment (philbucher) [reply kratos/includes/constitutive_law.h]: > Are we going to consider always 3D by default?

why not?
I mean this should have a size that can be used for everything

can you explain more your question?

**Review by philbucher — COMMENTED** [review]
maybe we should use `using` instead of `typedef` to be more C++11?
I think @sunethwarna knows a bit more, could you please comment  @sunethwarna ?

File: kratos/includes/constitutive_law.h
Comment (AlejandroCornejo) [reply kratos/includes/constitutive_law.h]: I agree with @philbucher , the obj is to use the max size andthen resize if it is lower dimension

Comment (loumalouomega) [reply kratos/includes/constitutive_law.h]: That was the question

File: kratos/includes/constitutive_law.h
Comment (RiccardoRossi) [reply kratos/includes/constitutive_law.h]: to be Honest i don't love this name.

what about "DeformationGradientType"

File: kratos/includes/constitutive_law.h
Comment (RiccardoRossi) [reply kratos/includes/constitutive_law.h]: i would not use the 3D, since it can be used with size 3,4 or 6

File: kratos/includes/constitutive_law.h
Comment (RiccardoRossi) [reply kratos/includes/constitutive_law.h]: unfortunately array_1d is actually of fixed size. We need to use "BoundedVector" to be resizeable.

in any case i would propose to do this change in two steps:
1. add the typedefs but with "Vector" and "Matrix" (so still heap array) and change the code so that in the CL we never use the Vector and Matrix but rather always the typedefs
2. actually change the typedefs to the BoundedVector/Matrix

**Comment by AlejandroCornejo** [comment]
Another concern is what to do with the CalculateValues... Now I am forced to do a static_cast<Vector>(rValues.GetStressVector) in order to be consistent

**Comment by loumalouomega** [comment]
> Another concern is what to do with the `CalculateValues`... Now I am forced to do a` static_cast<Vector>(rValues.GetStressVector)` in order to be consistent

I would update those methods as well, with the pain that involves...

File: kratos/includes/constitutive_law.h
Comment (rubenzorrilla) [reply kratos/includes/constitutive_law.h]: > 
> 
> > Another concern is what to do with the `CalculateValues`... Now I am forced to do a` static_cast<Vector>(rValues.GetStressVector)` in order to be consistent
> 
> I would update those methods as well, with the pain that involves...

Agree.

Comment (rubenzorrilla) [reply kratos/includes/constitutive_law.h]: Indeed, and considering that this PR will be a pain in the a**, I'd follow @RiccardoRossi 's  apprach but doing 2 PRs, one for step 1 and one for step 2.
In PR 1 I'd also change the function signatures as @loumalouomega suggests below. Even though I'm afraid that this will not raise any compilation error, so we will miss some of them, I think will make things easier for 2nd PR.

Comment (AlejandroCornejo) [reply kratos/includes/constitutive_law.h]: If so, the `CAUCHY_STRESS_VECTOR` variable should be a `StressVectorType `and not a `Vector `anymore right?

Comment (loumalouomega) [reply kratos/includes/constitutive_law.h]: Indeed. Maybe we should discuss this in more detail

Comment (AlejandroCornejo) [reply kratos/includes/constitutive_law.h]: ok... I'll make in this pr then the typedef's with the current `Vector `and `MAtrix` but changing all the stuff in the element and CL

**Comment by RiccardoRossi** [comment]
> Another concern is what to do with the `CalculateValues`... Now I am forced to do a` static_cast<Vector>(rValues.GetStressVector)` in order to be consistent

this has to change... static cast in this context is just wrong

File: kratos/includes/constitutive_law.h
Comment (RiccardoRossi) [reply kratos/includes/constitutive_law.h]: yes, and merge it ASAP (i mean even before changing all the CLs) since this way people can contribute to it

File: kratos/includes/constitutive_law.h
Comment (RiccardoRossi) [inline kratos/includes/constitutive_law.h:231]: eventually the good way to do this is to use "span" for vectors and "mdspan" for Matrix, but i would leave it for the future

**Comment by AlejandroCornejo** [comment]
> 
> 
> > Another concern is what to do with the `CalculateValues`... Now I am forced to do a` static_cast<Vector>(rValues.GetStressVector)` in order to be consistent
> 
> this has to change... static cast in this context is just wrong

Also the `CAUCHY_STRESS_VECTOR ` variable has to be changed in the future then

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: this is not correct... beware, ou cannot statically cast an array_1d to a Vector!

File: kratos/includes/constitutive_law.h
Comment (RiccardoRossi) [inline kratos/includes/constitutive_law.h:399]: this is another class, although defined within the ConstitutiveLaw class.

i wonder if you should not use "ConstitutiveLaw::DeformationGradientMatrixType"

Comment (RiccardoRossi) [reply kratos/includes/constitutive_law.h:399]: (same in the next few entries)

**Review by RiccardoRossi — COMMENTED** [review]
to me this is fine, thx for the change.

Nevertheless please wait until tomorrow to merge in case anyone wants to comment

**Review by RiccardoRossi — DISMISSED** [review]

**Review by philbucher — CHANGES_REQUESTED** [review]
this has still changes to the apps that are incorrect
I thought this PR was abt adding the typedefs?

**Comment by RiccardoRossi** [comment]
> this has still changes to the apps that are incorrect
> I thought this PR was abt adding the typedefs?

as far as i see it, the only changes are in using some of the typedefs in the App, so i see it correct.
the changes in the Element base class are also purely cometic

**Comment by loumalouomega** [comment]
Can we benchmark the differences between the two versions?

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: I meant this for for example

**Comment by RiccardoRossi** [comment]
> Can we benchmark the differences between the two versions?

there is no difference as of this PR

once we substituted the typedef everywhere, we will be able to change it only in the base class, and eventually we will be able to benchmark

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: you are right i forgot about this one

File: applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp]: and this one

**Comment by AlejandroCornejo** [comment]
Let's wait a sec until I'm sure it is not breaking anything...

**Comment by philbucher** [comment]
> Let's wait a sec until I'm sure it is not breaking anything...

yes please run the validation tests in the structural!

other than that (and the static_cast) no issues from my side

**Review by philbucher — DISMISSED** [review]
:+1:

**Comment by AlejandroCornejo** [comment]
I've run the tests and it seems that there are two that are failing, IDK if due to this PR or not...

======================================================================
FAIL: test_execution (structural_mechanics_test_factory.ShellT3ThickNonLinearStaticTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\Kratos\applications\StructuralMechanicsApplication\tests\structural_mechanics_test_factory.py", line 72, in test_execution

    self.test.RunSolutionLoop()
  File "D:\Kratos\bin\Release\KratosMultiphysics\analysis_stage.py", line 67, in RunSolutionLoop
    self.FinalizeSolutionStep()
  File "D:\Kratos\bin\Release\KratosMultiphysics\analysis_stage.py", line 149, in FinalizeSolutionStep
    process.ExecuteFinalizeSolutionStep()
  File "D:\Kratos\bin\Release\KratosMultiphysics\from_json_check_result_process.py", line 54, in ExecuteFinalizeSolutionStep
    self.assertTrue(self.process.IsCorrectResult(), "Results do not coincide with the JSON reference\n" + self.process.GetErrorMessage())
AssertionError: False is not true : Results do not coincide with the JSON reference
Error checking for variable REACTION_Y, results:
Node 1 -206.5734!=-140.6796; rel_tol=1e-06, abs_tol=0.001



FAIL: test_execution (structural_mechanics_test_factory.ShellQ4ThinNonLinearStaticTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\Kratos\applications\StructuralMechanicsApplication\tests\structural_mechanics_test_factory.py", line 72, in test_execution

    self.test.RunSolutionLoop()
  File "D:\Kratos\bin\Release\KratosMultiphysics\analysis_stage.py", line 67, in RunSolutionLoop
    self.FinalizeSolutionStep()
  File "D:\Kratos\bin\Release\KratosMultiphysics\analysis_stage.py", line 149, in FinalizeSolutionStep
    process.ExecuteFinalizeSolutionStep()
  File "D:\Kratos\bin\Release\KratosMultiphysics\from_json_check_result_process.py", line 54, in ExecuteFinalizeSolutionStep
    self.assertTrue(self.process.IsCorrectResult(), "Results do not coincide with the JSON reference\n" + self.process.GetErrorMessage())
AssertionError: False is not true : Results do not coincide with the JSON reference
Error checking for variable REACTION_Y, results:
Node 1 -160.0308!=-138.1294; rel_tol=1e-06, abs_tol=0.001

**Comment by philbucher** [comment]
those failures are known, I should finally update the resutls (see #7469)
Sorry for the onconvenience

**Comment by AlejandroCornejo** [comment]
approve whenever you want :)

**Review by philbucher — APPROVED** [review]

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**
